Web Inspector: Timelines: support exporting and importing data from worker targets

dcrousso opened the pull request:
#### 5d6c6dd65e724fa1ccdbf390ceeee95dd6a89775
<pre>
Web Inspector: Timelines: support exporting and importing data from worker targets
<a href="https://bugs.webkit.org/show_bug.cgi?id=287738">https://bugs.webkit.org/show_bug.cgi?id=287738</a>
&lt;<a href="https://rdar.apple.com/problem/145330533">rdar://problem/145330533</a>&gt;

Reviewed by BJ Burg.

This will allow developers to share timeline recordings of `Worker` with others, just like what&apos;s already possible for the main execution context.

* Source/WebInspectorUI/UserInterface/Protocol/Target.js:
(WI.Target.prototype.exportData): Added.
* Source/WebInspectorUI/UserInterface/Models/ImportedTarget.js: Added.
(WI.ImportedTarget):
(WI.ImportedTarget.import):
(WI.ImportedTarget.prototype.exportData):
(WI.ImportedTarget.prototype.forIdentifier):
(WI.ImportedTarget.prototype.get identifier):
(WI.ImportedTarget.prototype.get type):
(WI.ImportedTarget.prototype.get name):
(WI.ImportedTarget.prototype.get url):
(WI.ImportedTarget.prototype.get isDestroyed):
(WI.ImportedTarget.prototype.get displayName):
(WI.ImportedTarget.prototype.hasDomain):
(WI.ImportedTarget.prototype.hasCommand):
(WI.ImportedTarget.prototype.hasEvent):
Add a model object to represent a target that did exist when the timeline recording was captured.

* Source/WebInspectorUI/UserInterface/Models/TimelineRecording.js:
(WI.TimelineRecording):
(WI.TimelineRecording.async import):
(WI.TimelineRecording.prototype.exportData):
(WI.TimelineRecording.prototype.reset):
(WI.TimelineRecording.prototype.updateCallingContextTrees):
* Source/WebInspectorUI/UserInterface/Models/HeapAllocationsTimelineRecord.js:
(WI.HeapAllocationsTimelineRecord.async fromJSON):
(WI.HeapAllocationsTimelineRecord.prototype.toJSON):
* Source/WebInspectorUI/UserInterface/Models/ScriptTimelineRecord.js:
(WI.ScriptTimelineRecord.async fromJSON):
(WI.ScriptTimelineRecord.prototype.toJSON):
Include the `WI.Target` data in the exported data to create a `WI.ImportedTarget` when (re)imported.

* Source/WebInspectorUI/UserInterface/Models/ScriptTimeline.js:
(WI.ScriptTimeline.prototype.get imported): Added.
Provide a way for `WI.ScriptClusterTimelineView` to know if a timeline recording is imported since it doesn&apos;t have access to the `WI.TimelineRecording`.

* Source/WebInspectorUI/UserInterface/Proxies/HeapSnapshotWorkerProxy.js:
(WI.HeapSnapshotWorkerProxy.prototype.createImportedSnapshot):
* Source/WebInspectorUI/UserInterface/Views/HeapAllocationsTimelineView.js:
(WI.HeapAllocationsTimelineView.prototype._importButtonNavigationItemClicked):
Remove the FIXME comment since imported snapshots will have a `null` target in the `HeapSnapshotWorker` since the `WI.Target` (and `WI.ImportedTarget`) only exist in the main execution context.

* Source/WebInspectorUI/UserInterface/Views/HeapSnapshotDataGridTree.js:
(WI.HeapSnapshotObjectGraphDataGridTree.prototype.populateTopLevel):
Just query for all kinds of global objects instead of deciding based on the `WI.TargetType`.

* Source/WebInspectorUI/UserInterface/Proxies/HeapSnapshotProxy.js:
(WI.HeapSnapshotProxy):
(WI.HeapSnapshotProxy.prototype.get imported):
* Source/WebInspectorUI/UserInterface/Proxies/HeapSnapshotNodeProxy.js:
(WI.HeapSnapshotNodeProxy):
(WI.HeapSnapshotNodeProxy.prototype.get imported): Added.
* Source/WebInspectorUI/UserInterface/Views/HeapSnapshotInstanceDataGridNode.js:
(WI.HeapSnapshotInstanceDataGridNode.logHeapSnapshotNode):
(WI.HeapSnapshotInstanceDataGridNode.prototype.createCellContent):
(WI.HeapSnapshotInstanceDataGridNode.prototype._getRemoteObject):
(WI.HeapSnapshotInstanceDataGridNode.prototype._populateWindowPreview):
(WI.HeapSnapshotInstanceDataGridNode.prototype._populatePreview):
(WI.HeapSnapshotInstanceDataGridNode.prototype._populateRemoteObject):
(WI.HeapSnapshotInstanceDataGridNode.prototype._mouseoverHandler):
Prevent imported heap snaphots from using `Heap` commands.

* Source/WebInspectorUI/UserInterface/Views/ScriptClusterTimelineView.js:
(WI.ScriptClusterTimelineView):
(WI.ScriptClusterTimelineView.prototype.restoreFromCookie):
(WI.ScriptClusterTimelineView.prototype._handleTargetAdded):
Only default to the `WI.mainTarget` if the timeline recording is not imported.

* Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js:
(WI.DebuggerManager.prototype.scriptForIdentifier):
(WI.DebuggerManager.prototype.scriptsForURL):
Drive-by: Add a fastpath for `WI.ImportedTarget`.

* Source/WebInspectorUI/UserInterface/Views/ProfileDataGridNode.js:
(WI.ProfileDataGridNode.prototype.iconClassName):
(WI.ProfileDataGridNode.prototype.filterableDataForColumn):
(WI.ProfileDataGridNode.prototype._getScript): Added.
(WI.ProfileDataGridNode.prototype._displayContent):
Drive-by: Attempt to get the `WI.Script` (or `WI.Resource`) via the URL if viewing an imported timeline recording so that if the developer has imported the recording onto the same website then they can see source code locations.

* Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js:
(HeapSnapshot.prototype.get imported): Added.
* Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshotWorker.js:
(HeapSnaphotWorker.prototype.createSnapshot):
Drive-by: Add missing method now that `imported` is no longer given as a separate argument.

* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/UserInterface/Test.html:

* LayoutTests/inspector/timeline/timeline-recording.html:
* LayoutTests/inspector/timeline/timeline-recording-expected.txt:

Canonical link: <a href="https://commits.webkit.org/292991@main">https://commits.webkit.org/292991@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0b8e60c19990c6a818b5eb87cbe2e45ca64b6a95

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97628 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17253 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7469 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102734 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48157 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17547 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25706 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/74373 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31555 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100631 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13292 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/88265 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54722 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13064 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6160 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47599 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/83085 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6236 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104735 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24708 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/18020 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/83420 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25080 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84402 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82849 "Found 1 new API test failure: /WebKitGTK/TestCookieManager:/webkit/WebKitCookieManager/persistent-storage (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20860 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27403 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/5084 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18303 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24669 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/29838 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24491 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27805 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26065 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

burg commented on Source/WebInspectorUI/UserInterface/Models/ImportedTarget.js:
```diff
@@ -0,0 +1,101 @@
+    }
+};
+
+WI.ImportedTarget._forIdentifier = new Map;
```
Nit: since the static method is named nicely, can this be `_forIdentifierMap` ?

burg commented on Source/WebInspectorUI/UserInterface/Models/TimelineRecording.js:
```diff
@@ -85,21 +81,21 @@ WI.TimelineRecording = class TimelineRecording extends WI.Object
+        if (samples) {
+            for (let {target, stackTraces, durations} of samples)
+                recording.updateCallingContextTrees(WI.ImportedTarget.import(target), stackTraces, durations);
+        } else if (json.sampleStackTraces && json.sampleDurations)
```
Add comment about using these keys for compatibility with older saved exported timeline recordings.

dcrousso committed:
Web Inspector: Timelines: support exporting and importing data from worker targets
https://bugs.webkit.org/show_bug.cgi?id=287738
<rdar://problem/145330533>

Reviewed by BJ Burg.

This will allow developers to share timeline recordings of `Worker` with others, just like what's already possible for the main execution context.

* Source/WebInspectorUI/UserInterface/Protocol/Target.js:
(WI.Target.prototype.exportData): Added.
* Source/WebInspectorUI/UserInterface/Models/ImportedTarget.js: Added.
(WI.ImportedTarget):
(WI.ImportedTarget.import):
(WI.ImportedTarget.prototype.exportData):
(WI.ImportedTarget.prototype.forIdentifier):
(WI.ImportedTarget.prototype.get identifier):
(WI.ImportedTarget.prototype.get type):
(WI.ImportedTarget.prototype.get name):
(WI.ImportedTarget.prototype.get url):
(WI.ImportedTarget.prototype.get isDestroyed):
(WI.ImportedTarget.prototype.get displayName):
(WI.ImportedTarget.prototype.hasDomain):
(WI.ImportedTarget.prototype.hasCommand):
(WI.ImportedTarget.prototype.hasEvent):
Add a model object to represent a target that did exist when the timeline recording was captured.

* Source/WebInspectorUI/UserInterface/Models/TimelineRecording.js:
(WI.TimelineRecording):
(WI.TimelineRecording.async import):
(WI.TimelineRecording.prototype.exportData):
(WI.TimelineRecording.prototype.reset):
(WI.TimelineRecording.prototype.updateCallingContextTrees):
* Source/WebInspectorUI/UserInterface/Models/HeapAllocationsTimelineRecord.js:
(WI.HeapAllocationsTimelineRecord.async fromJSON):
(WI.HeapAllocationsTimelineRecord.prototype.toJSON):
* Source/WebInspectorUI/UserInterface/Models/ScriptTimelineRecord.js:
(WI.ScriptTimelineRecord.async fromJSON):
(WI.ScriptTimelineRecord.prototype.toJSON):
Include the `WI.Target` data in the exported data to create a `WI.ImportedTarget` when (re)imported.

* Source/WebInspectorUI/UserInterface/Models/ScriptTimeline.js:
(WI.ScriptTimeline.prototype.get imported): Added.
Provide a way for `WI.ScriptClusterTimelineView` to know if a timeline recording is imported since it doesn't have access to the `WI.TimelineRecording`.

* Source/WebInspectorUI/UserInterface/Proxies/HeapSnapshotWorkerProxy.js:
(WI.HeapSnapshotWorkerProxy.prototype.createImportedSnapshot):
* Source/WebInspectorUI/UserInterface/Views/HeapAllocationsTimelineView.js:
(WI.HeapAllocationsTimelineView.prototype._importButtonNavigationItemClicked):
Remove the FIXME comment since imported snapshots will have a `null` target in the `HeapSnapshotWorker` since the `WI.Target` (and `WI.ImportedTarget`) only exist in the main execution context.

* Source/WebInspectorUI/UserInterface/Views/HeapSnapshotDataGridTree.js:
(WI.HeapSnapshotObjectGraphDataGridTree.prototype.populateTopLevel):
Just query for all kinds of global objects instead of deciding based on the `WI.TargetType`.

* Source/WebInspectorUI/UserInterface/Proxies/HeapSnapshotProxy.js:
(WI.HeapSnapshotProxy):
(WI.HeapSnapshotProxy.prototype.get imported):
* Source/WebInspectorUI/UserInterface/Proxies/HeapSnapshotNodeProxy.js:
(WI.HeapSnapshotNodeProxy):
(WI.HeapSnapshotNodeProxy.prototype.get imported): Added.
* Source/WebInspectorUI/UserInterface/Views/HeapSnapshotInstanceDataGridNode.js:
(WI.HeapSnapshotInstanceDataGridNode.logHeapSnapshotNode):
(WI.HeapSnapshotInstanceDataGridNode.prototype.createCellContent):
(WI.HeapSnapshotInstanceDataGridNode.prototype._getRemoteObject):
(WI.HeapSnapshotInstanceDataGridNode.prototype._populateWindowPreview):
(WI.HeapSnapshotInstanceDataGridNode.prototype._populatePreview):
(WI.HeapSnapshotInstanceDataGridNode.prototype._populateRemoteObject):
(WI.HeapSnapshotInstanceDataGridNode.prototype._mouseoverHandler):
Prevent imported heap snaphots from using `Heap` commands.

* Source/WebInspectorUI/UserInterface/Views/ScriptClusterTimelineView.js:
(WI.ScriptClusterTimelineView):
(WI.ScriptClusterTimelineView.prototype.restoreFromCookie):
(WI.ScriptClusterTimelineView.prototype._handleTargetAdded):
Only default to the `WI.mainTarget` if the timeline recording is not imported.

* Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js:
(WI.DebuggerManager.prototype.scriptForIdentifier):
(WI.DebuggerManager.prototype.scriptsForURL):
Drive-by: Add a fastpath for `WI.ImportedTarget`.

* Source/WebInspectorUI/UserInterface/Views/ProfileDataGridNode.js:
(WI.ProfileDataGridNode.prototype.iconClassName):
(WI.ProfileDataGridNode.prototype.filterableDataForColumn):
(WI.ProfileDataGridNode.prototype._getScript): Added.
(WI.ProfileDataGridNode.prototype._displayContent):
Drive-by: Attempt to get the `WI.Script` (or `WI.Resource`) via the URL if viewing an imported timeline recording so that if the developer has imported the recording onto the same website then they can see source code locations.

* Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js:
(HeapSnapshot.prototype.get imported): Added.
* Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshotWorker.js:
(HeapSnaphotWorker.prototype.createSnapshot):
Drive-by: Add missing method now that `imported` is no longer given as a separate argument.

* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/UserInterface/Test.html:

* LayoutTests/inspector/timeline/timeline-recording.html:
* LayoutTests/inspector/timeline/timeline-recording-expected.txt:

Canonical link: https://commits.webkit.org/292991@main
